### PR TITLE
Easily available custom package server

### PR DESCRIPTION
**Pull request by changlinli**
**Quick Summary:**

Creating a new custom package repository server that decouples us from GitHub and git and serves as the basis for a separate online service that can offer one-click custom package repositories. See `CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md` for more details.

### CODE REVIEW

File: CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+   POST request to the `/register` endpoint.
+
+Note that while it is not necessary to discard `git` and GitHub to adhere to
+this specification, the ability to directly upload zipfiles along with a
+separate system for authentication does obviate their need.
+
+Otherwise, we'll follow the standard Elm package server API pretty closely, so
+our overall API endpoints are the same as the standard Elm package server.
+
+As a point of reference, we'll first the standard Elm package server API
```
Comment (supermario) [inline CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md]:
```suggestion
As a point of reference, we'll first cover the standard Elm package server API
```

File: CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+complicates backups. There are solutions that allow for us to live-replicate a
+DB to ensure we never lose any data. Those solutions seem to be a bit trickier
+for filesystems (most I've seen seem to replicate on some predetermined
+schedule).
+
+In order to do per-user authentication (something we'll need to implement since
+we're dropping the GitHub dependence), we'll need some sort of database anyway.
+
+As such, I've decided to code up the package server from scratch. There are few
+enough API endpoints that it's a feasible task and the moving away from . The package
```
Comment (supermario) [inline CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md]: Finish this thought? :)

**Review by supermario — APPROVED** [review]
This is cool – but yeah a lot to review 😅 

Some drive-by thoughts:

- SQLite as an embedded and portable DB could be worth considering rather than hand rolling another on-disk format (your design notes were incomplete so unsure if you did further exploration of this)
- You might consider extracting more of the modifications to the core Elm files if you haven't already, to keep future potential diffs against upstream cleaner. They don't look particularly large, and obviously diminishing returns when added functions need to use local types, but I notice you're mixing in imports and functions – having a clear `-- @zokka` marker where additions are clearly appended to the ends of the imports / file may help maintenance (and clarity on what's zokka divergence and what's not) in future.

I've not had a chance to try this out yet, but hope to check it out this weekend 👍

File: CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md (hunk 2 above)
Comment (changlinli) [reply CUSTOM_PACKAGE_SERVER_REFERENCE_NOTES.md]: Ah yes, done!

**Comment by changlinli** [comment]
Yep @supermario you're right on the money! I decided to go with SQLite instead.

I agree with you that I should move out all the Zokka stuff to make it more clear what is new in Zokka and what is in Elm. I'll likely do that as a separate PR to combine with the pre-existing Zokka additions I have.

**Pull request merged**